clippy_lints/src/methods/obfuscated_if_else.rs

```diff
@@ -11,31 +13,51 @@ pub(super) fn check<'tcx>(
     then_recv: &'tcx hir::Expr<'_>,
     then_arg: &'tcx hir::Expr<'_>,
     unwrap_arg: &'tcx hir::Expr<'_>,
+    then_method_name: &str,
 ) {
-    // something.then_some(blah).unwrap_or(blah)
-    // ^^^^^^^^^-then_recv ^^^^-then_arg   ^^^^- unwrap_arg
-    // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- expr
-
     let recv_ty = cx.typeck_results().expr_ty(then_recv);
 
     if recv_ty.is_bool() {
         let mut applicability = Applicability::MachineApplicable;
+        let if_then = match then_method_name {
+            "then" if let ExprKind::Closure(closure) = then_arg.kind => {
+                let body = cx.tcx.hir().body(closure.body);
+                snippet(cx, body.value.span, "..")
+            },
+            "then_some" => snippet_with_applicability(cx, then_arg.span, "..", &mut applicability),
+            _ => String::new().into(),
+        };
+
         let sugg = format!(
             "if {} {{ {} }} else {{ {} }}",
-            snippet_with_applicability(cx, then_recv.span, "..", &mut applicability),
-            snippet_with_applicability(cx, then_arg.span, "..", &mut applicability),
+            remove_needless_parentheses(&snippet_with_applicability(
+                cx,
+                then_recv.span,
+                "..",
+                &mut applicability
+            )),
+            if_then,
             snippet_with_applicability(cx, unwrap_arg.span, "..", &mut applicability)
         );
 
         span_lint_and_sugg(
             cx,
             OBFUSCATED_IF_ELSE,
             expr.span,
-            "use of `.then_some(..).unwrap_or(..)` can be written \
-            more clearly with `if .. else ..`",
+            "use of method chains can be written more clearly with `if .. else ..`",
             "try",
             sugg,
             applicability,
         );
     }
 }
+
+// Remove the enclosing parentheses if they exist
+// e.g. "(a == 1)" -> "a == 1"
+fn remove_needless_parentheses(s: &str) -> String {
+    if has_enclosing_paren(s) {
+        s[1..s.len() - 1].to_string()
+    } else {
```

Comment:
Even if you didn't use `Sugg`, you could return a `&str` instead of allocating a `String`